gax/src/main/java/com/google/api/gax/tracing/OpencensusTracer.java

```diff
@@ -0,0 +1,386 @@
+
+    span.putAttributes(attributes);
+    span.end(EndSpanOptions.builder().setStatus(convertErrorToStatus(error)).build());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void connectionSelected(int id) {
+    span.addAnnotation(
+        "Connection selected", ImmutableMap.of("id", AttributeValue.longAttributeValue(id)));
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    currentAttemptId = attemptNumber;
+    attemptRequests = 0;
+    attemptResponses = 0;
+
+    HashMap<String, AttributeValue> attributes = Maps.newHashMap();
+    populateAttemptNumber(attributes);
+
+    span.addAnnotation("Attempt started", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptSucceeded() {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+
+    span.addAnnotation("Attempt succeeded", attributes);
+  }
+
+  @Override
+  public void attemptCancelled() {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+
+    span.addAnnotation("Attempt cancelled", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptFailed(Throwable error, Duration delay) {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    attributes.put("delay ms", AttributeValue.longAttributeValue(delay.toMillis()));
+    populateError(attributes, error);
+
+    String msg = error != null ? "Attempt failed" : "Operation incomplete";
+    span.addAnnotation(msg + ", scheduling next attempt", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptFailedRetriesExhausted(Throwable error) {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    populateError(attributes, error);
+
+    span.addAnnotation("Attempts exhausted", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void attemptPermanentFailure(Throwable error) {
+    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    populateError(attributes, error);
+
+    span.addAnnotation("Attempt failed, error not retryable", attributes);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void responseReceived() {
+    attemptResponses++;
+    totalAttemptResponses++;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void requestSent() {
+    attemptRequests++;
+    totalAttemptRequests++;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void batchRequestSent(long elementCount, long requestSize) {
+    span.putAttribute("batch count", AttributeValue.longAttributeValue(elementCount));
+    span.putAttribute("request size", AttributeValue.longAttributeValue(requestSize));
+  }
+
+  private Map<String, AttributeValue> baseOperationAttributes() {
+    HashMap<String, AttributeValue> attributes = Maps.newHashMap();
+
+    attributes.put("attempt count", AttributeValue.longAttributeValue(currentAttemptId + 1));
+
+    if (totalAttemptRequests > 0) {
```

Comment:
It's to avoid generating noise for unary requests and single item streams. Also note that this isn't counting attempts but messages sent. I'll rename the variables to *SentMessages and *ReceivedMessages to avoid confusion in the future